nesta/core/orms/orm_utils.py

```diff
@@ -428,14 +453,116 @@ def _filter_out_duplicates(session, Base, _class, data,
+        all_rows_stmt = key.in_(tuple(pk for pk, in pks))
+    delete_stmt = _class.__table__.delete().where(all_rows_stmt)
+    return delete_stmt
+
+
+def is_null(x):
+    """Wrapper around pandas isnull with an exception for iterables,
+    which cause a ValueError on conversion to boolean."""
+    try:
+        return bool(pd.isnull(x))
+    except ValueError:
+        return False
+
+
+def merge_duplicates(db_env, section, database,
+                     Base, _class, data, low_memory):
+    """Alternative to `filter_out_duplicates`: Find all duplicates in the list of data,
+    and also for duplicates between the provided data and the database, and then
+    select the most recent non-null value for upsertion.
+
+    Args:
+        session (:obj:`sqlalchemy.orm.session.Session`): SqlAlchemy session object.
+        Base (:obj:`sqlalchemy.Base`): The Base ORM for this data.
+        _class (:obj:`sqlalchemy.Base`): The ORM for this data.
+        data (:obj:`list` of :obj:`dict`): Rows of data to insert
+        low_memory (bool): If the pkeys are few or small types (i.e. they won't
+                           occupy lots of memory) then set this to True.
+                           This will speed things up significantly (like x 100),
+                           but will blow up for heavy pkeys or large tables.
+    Returns:
+        :obj:`list` of :obj:`_class` instantiated by data, with duplicate pks removed.
+    """
+    pkey_cols = _class.__table__.primary_key.columns
+    is_auto_pkey = check_is_auto_pkey(pkey_cols)
+    if is_auto_pkey:
+        raise ValueError('AutoPK fields cannot be merged, you must set smart_update = False')
+    if not low_memory:
+        raise NotImplementedError('low_memory mode has not been implemented for `merge_duplicates`.'
+                                  'Use smart_update = False')
+    # Group objects into sets of duplicates
+    session = get_session(db_env, section, database, Base)
+    all_pks = (get_all_pks(session, _class, pkey_cols) if low_memory else set())
+    pks_to_drop = []
+    pk_row_lookup = defaultdict(list)
+    for row in data:
+        pk = generate_pk(row, pkey_cols)
+        pk_row_lookup[pk].append(row)
+        if pk in all_pks:
+            _row = retrieve_row_by_pk(session, _class, row, pkey_cols)
+            pk_row_lookup[pk].append(_row)
+            pks_to_drop.append(pk)
+    delete_stmt = create_delete_stmt(_class, pkey_cols, pks_to_drop)
+    session.close()
+
+    # Now merge the fields by taking the first non-null value
+    objs = []
+    for pk, rows in pk_row_lookup.items():
+        merged_row = {}
+        for col in rows[0].keys():
+            value = None
+            for row in rows:
+                if not is_null(row[col]):
+                    value = row[col]
+                    break
+            merged_row[col] = value
+        #objs.append(_class(**merged_row))
+        objs.append(merged_row)
+    return objs, delete_stmt, None
```

Comment:
Not reviewed this bundle of functions yet, as they're a little hard to follow (Lose track of types not being familiar with SQLalchemy etc.) and there's no tests to help.